src/main/scala/mesosphere/marathon/api/TaskKiller.scala

```diff
@@ -34,32 +36,40 @@ class TaskKiller @Inject() (
       case Some(app) =>
         checkAuthorization(UpdateRunSpec, app)
 
+        // TODO: We probably want to pass the execution context as an implcit.
         import scala.concurrent.ExecutionContext.Implicits.global
-        taskTracker.appTasks(appId).flatMap { allTasks =>
+
+        async {
+          val allTasks = await(taskTracker.appTasks(appId))
           val foundTasks = findToKill(allTasks)
-          val expungeTasks = if (wipe) expunge(foundTasks) else Future.successful(())
-
-          expungeTasks.map { _ =>
-            val launchedTasks = foundTasks.filter(_.launched.isDefined)
-            if (launchedTasks.nonEmpty) {
-              service.killTasks(appId, launchedTasks)
-              foundTasks
-            } else foundTasks
-          }
+
+          if (wipe) expunge(foundTasks)
+
+          val launchedTasks = foundTasks.filter(_.launched.isDefined)
+          if (launchedTasks.nonEmpty) await(service.killTasks(appId, launchedTasks))
+          // TODO: For some reasons tasks returned by service.killTasks !=
+          // foundTasks in some cases. What should be returned?
+          foundTasks
         }
 
       case None => Future.failed(UnknownAppException(appId))
     }
   }
 
-  private[this] def expunge(tasks: Iterable[Task])(implicit ec: ExecutionContext): Future[Unit] = {
-    tasks.foldLeft(Future.successful(())) { (resultSoFar, nextTask) =>
-      resultSoFar.flatMap { _ =>
+  private[this] def expunge(tasks: Iterable[Task])(implicit ec: ExecutionContext): Unit = {
+    tasks.foreach { nextTask =>
+      async {
         log.info("Expunging {}", nextTask.taskId)
-        stateOpProcessor.process(TaskStateOp.ForceExpunge(nextTask.taskId)).map(_ => ()).recover {
-          case NonFatal(cause) =>
-            log.info("Failed to expunge {}, got: {}", Array[Object](nextTask.taskId, cause): _*)
-        }
+
+        val processNextTask = stateOpProcessor.process(TaskStateOp.ForceExpunge(nextTask.taskId))
+          .map(_ => Done)
+          .recover {
+            case NonFatal(cause) =>
+              log.info("Failed to expunge {}, got: {}", Array[Object](nextTask.taskId, cause): _*)
+              Done // Note, this error is swallowed.
+          }
+
+        await(processNextTask)
```

Comment:
This is supposed to make it more explicit that we purge sequentially and not in parallel.